### DISCUSSION TITLE
Lanauge Server Extension

### DISCUSSION THREAD

**Discussion by Quafadas** [disneystreaming/smithy4s#354] [Q&A] [answered]
I've added the VSCode extension for the smithy LSP server. 

Unfortuntely, I haven't gottent it to work yet. I suspect, that may be because I work in a borderline absurd environment that I have little control over, but I'd like to check if it's something easier :-)! Here's what I can find. 

In the Extension Host Log; 
```
[2022-08-04 16:10:37.821] [exthost] [info] ExtensionService#_doActivateExtension disneystreaming.smithy, startup: false, activationEvent: 'onLanguage:smithy'
```
So it looks like it's registered. In .smithy-build.json
```
{    
    "mavenDependencies": [
        "com.disneystreaming.smithy4s:smithy4s-protocol_2.13:latest.stable"
    ],
    "mavenRepositories": ["https://packages.corp.com/artifactory/maven"]
}
```
In the extension window in VSCode itself, there is a Runtime Status tab. In that tab, I see

```
Activation time : 162ms
------
Uncaught Errors (1)
spawn UNKNOWN
```
Which looks like a hint that something has gone wrong. 

I haven't (yet) found any other logging... any obvious hints / avenues of investigation?

P.S. For the record, I tried latest.stable with a conrete version, wasnt' sure it it was a wilcard or not ... but same behaviour...

**Comment by Baccata** [comment, marked as answer]
@Quafadas it's very likely not gonna be a smithy4s issue, so let's put the configuration aside for now. 

When you say "haven't got it to work", what does it mean exactly ? What doesn't work ? Do you get syntax colouring but none of the other features ? Does nothing happen at all when you open a smithy file ? 

Can you try the [AWS-owned vscode extension](https://marketplace.visualstudio.com/items?itemName=Smithy.smithy-vscode-extension) ? We've actually collaborated with them on the LSP implementation, and only released our own vscode plugin because timeframes, but it's likely that we'll deprecate it in favour of theirs at some point in the future.

**Reply by Quafadas** [reply]
So I get sytax highlighting, purple annotations, green strings, yellow names. No goto definition or autocomplete... 

I'll try the AWS extension.. thanks, as I say, I suspect the problem really lies eleswhere

**Reply by Baccata** [reply]
You could potentially check-out the extension code, and run it in debug mode (vscode makes it really easy). 

Another set of questions : 

* What OS are you running ? 
* What JVM version are you running ?
* Have you got [coursier](https://github.com/coursier/coursier) installed ? in theory the extension should download it by itself if you don't have it installed
* The problem is likely to be coming from [there](https://github.com/disneystreaming/vscode-smithy/blob/main/src/extension.ts#L65-L68), so it could be that the extension fails to download coursier for whatever reason. You may want to try installing it manually

**Reply by Quafadas** [reply]
I had a dig into this and it's the coursier installation part. This kind of goes back to my original assumption about the environment I work in. 

1. Coursier is not installed on the path. As far as I am aware, it is not allowed to be in my org. 
2. Coursier CLI is not allowed to be downloaded - such URLs are blocked by default. 

I didn't dig further. I'm assuming that will generate an unrecoverable exception / situation. 

Conclusion: This is not a problem for either you or me (sadly). It is what it is - my apologies for the time taken.

**Reply by Baccata** [reply]
No problem, I understand the reality of the industry.

You might be able to build coursier from sources if you want to keep pursuing this.

**Reply by daddykotex** [reply]
Would it be worth having more information when setup fails (i.e. downloading `cs` if it's not on the path)?

**Reply by Baccata** [reply]
Yeah it would certainly be good to have error popups when that's the case.

**Reply by daddykotex** [reply]
My PR at https://github.com/disneystreaming/vscode-smithy/pull/10
Would certainly improve things. The latest commits that I've introduced are just validation that everything works but they don't do anything to address this issue. This commit does that: https://github.com/disneystreaming/vscode-smithy/pull/10/commits/c0b22e47ce7ef2ac9ead41959c30fb8f27bd3c57

Hopefully when merged and released, it can provide a better exp:
![Screen Shot 2022-08-09 at 12 44 00](https://user-images.githubusercontent.com/5230460/183742184-4a811ad1-3353-4e3b-8670-75272378b27e.png)
